pulser-simulation/pulser_simulation/hamiltonian.py

```diff
@@ -0,0 +1,640 @@
+# Copyright 2023 Pulser Development Team
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Defines the Hamiltonian class."""
+
+from __future__ import annotations
+
+import itertools
+import warnings
+from collections import defaultdict
+from collections.abc import Mapping
+from dataclasses import asdict
+from typing import Any, Union, cast
+
+import numpy as np
+import qutip
+
+from pulser.devices._device_datacls import BaseDevice
+from pulser.register.base_register import QubitId
+from pulser.sampler.samples import SequenceSamples, _PulseTargetSlot
+from pulser_simulation.simconfig import SimConfig
+
+
+class Hamiltonian:
+    r"""Generates Hamiltonian from a sampled sequence and noise.
+
+    Args:
+        samples_obj: A sampled sequence whose ChannelSamples have same
+            duration.
+        qdict: A dictionary associating coordinates to qubit ids.
+        device: The device specifications.
+        sampling_rate: The fraction of samples that we wish to extract from
+            the samples to simulate. Has to be a value between 0.05 and 1.0.
+        config: Configuration to be used for this simulation.
```

Comment:
Ideally for me, ` Hamiltonian` wouldn't deal with `SimConfig` that has irrelevant info like `runs`, `samples_per_run` and `solver_options`. Instead, it would deal only with `NoiseModel`, which can be easily obtained from a `SimConfig` using `SimConfig.to_noise_model()`.

This is something that can also be tackled in a second PR though